server/src/main/java/io/druid/segment/realtime/appenderator/Appenderator.java

```diff
@@ -61,8 +61,6 @@
    * This method may trigger a {@link #persistAll(Committer)} using the supplied Committer. If it does this, the
    * Committer is guaranteed to be *created* synchronously with the call to add, but will actually be used
    * asynchronously.
-   * <p>
-   * The add, clear, persist, persistAll, and push methods should all be called from the same thread.
```

Comment:
> I think synchronization is not enough because we need to guarantee that the execution order matches the method call order. Otherwise, we need a kind of sorted queue which provides such a guarantee. I think the easiest way is enforcing the methods are called by the same thread as before.

Yeah, I think that if it's feasible to require that the methods are called by the same thread, then that's simplest from the point of view of the Appenderator implementation.

> I think it makes sense to add a new push() method which takes a max retry count.

You mean `Appenderator.push()`, and having that method do its own retries? That makes sense to me. The ability to do retries in the Appenderator certainly makes life easier for its callers.

It sounds like it also makes it easier to ensure that the Committer metadata is always in sync with data written to disk, which is essential for correct behavior.